### ISSUE TITLE
DDTD -> Refactoring the responses. Let's start with cancelOrder

### ISSUE THREAD

**Issue by programarivm** [ExchangeUnion/xud#322]
Let me start by saying that I wonder if this issue sounds right.

As explained at [Status codes and their use in gRPC](https://grpc.io/grpc/core/md_doc_statuscodes.html):

> All RPCs started at a client return a status object composed of an integer code and a string message. The server-side can choose the status it returns for a given RPC.

The example available at https://github.com/grpc/grpc/issues/9053 shows this idea:

```
{
  code: 5,
  message: 'Operation operations/doesNotExist does not exist'
}
```
The objective is to build responses like these:

```
{ 
  "code": 0, 
  "message": "The order was cancelled successfully."
}

{
  "code": 3, 
  "message": "The order contains invalid arguments."
}

{
  "code": 5, 
  "message":"The order does not exist."
}
```
In such a way that the tests available at this moment will pass -- look at this [example](https://github.com/programarivm/xud/blob/ddtd-framework/test/ddtd/grpc/cancelOrder/GrpcStatus0.spec.ts):

```
client.cancelOrder(request, (response) => {
  assert.equal(0, response.toObject().code);
});
```

**Comment by programarivm** [comment]
More specifically, if this is right:

> All RPCs started at a client return a status object composed of an integer code and a string message. The server-side can choose the status it returns for a given RPC.

From a client perspective we should have something like this:

```
client.cancelOrder(request, (status) => {
  assert.equal(0, status.toObject().code);
});
```

**Comment by sangaman** [comment]
Yes we should definitely try to have the proper error code and a descriptive message on errors. On success, I don't believe we are specifying a particular message or code (although grpc might set code 0 automatically), we just return the response.

Is the test you cited failing? I haven't tried it yet.

**Comment by programarivm** [comment]
Please correct me if wrong, but currently if there's an error, the grpc server returns a null response. So, I just felt free to suggest this new interface using a `Status` object to literally:

> Return a status object composed of an integer code and a string message.

It could be like this:

```
{ 
  "code": 0, 
  "message": "The order was cancelled successfully.",
  "details": ...
}

{
  "code": 3, 
  "message": "The order contains invalid arguments."
}

{
  "code": 5, 
  "message":"The order does not exist."
}
```

The details can be used when necessary.

Note that a Status-based interface is a game changer in a sense that the common norm for the client uses an error with a message (or an error with a response) as described at [gRPC Errors](http://avi.im/grpc-errors/).

However, IMHO, using an error with a response is a bit semantically misleading because not all grpc status codes are errors. This way, you're encouraged to throw errors from minute one -- which by definition are the opposite of a response. And also, it is a XOR (exclusive or) way to solve the problem where you cannot have an error with a response together at the same time.

I guess this is why the server is currently returning null responses when errors occur, because xud is sticking with a XOR way to handle the errors.

Sorry but I am still learning about grpc. At this moment I am just wondering what is wrong with a grpc client using a `Status` object only?

**Comment by sangaman** [comment]
When there's an error the grpc service returns an error code and message. Or do you mean adding `code` and `message` fields to all of our grpc responses, rather than throwing errors? I haven't seen that with any other grpc services.

I think it's fine that an error won't have a response body or anything. We can append metadata to the response if we need to return additional information besides the type of error and message.

**Comment by sangaman** [comment]
I've just been trying to follow the examples and documentation for grpc services, which all seem to follow this pattern of null responses on errors. I don't think we should deviate from that unless we have a very good reason to.
